api/src/permissions/utils/fetch-dynamic-variable-context.ts

```diff
@@ -33,36 +24,93 @@ export async function _fetchDynamicVariableContext(options: FetchDynamicVariable
-		const usersService = new UsersService(context);
+		contextData['$CURRENT_USER'] = await fetchContextData(
+			'$CURRENT_USER',
+			permissionContext,
+			{ user: options.accountability.user },
+			async (fields) => {
+				const usersService = new UsersService(context);
 
-		contextData['$CURRENT_USER'] = await usersService.readOne(options.accountability.user, {
-			fields: Array.from(permissionContext.$CURRENT_USER!),
-		});
+				return await usersService.readOne(options.accountability.user!, {
+					fields,
+				});
+			},
+		);
 	}
 
 	if (options.accountability.role && (permissionContext.$CURRENT_ROLE?.size ?? 0) > 0) {
-		const rolesService = new RolesService(context);
+		contextData['$CURRENT_ROLE'] = await fetchContextData(
+			'$CURRENT_ROLE',
+			permissionContext,
+			{ role: options.accountability.role },
+			async (fields) => {
+				const usersService = new RolesService(context);
 
-		contextData['$CURRENT_ROLE'] = await rolesService.readOne(options.accountability.role, {
-			fields: Array.from(permissionContext.$CURRENT_ROLE!),
-		});
+				return await usersService.readOne(options.accountability.role!, {
+					fields,
+				});
+			},
+		);
 	}
 
 	if (options.accountability.roles.length > 0 && (permissionContext.$CURRENT_ROLES?.size ?? 0) > 0) {
-		const rolesService = new RolesService(context);
+		contextData['$CURRENT_ROLES'] = await fetchContextData(
+			'$CURRENT_ROLES',
+			permissionContext,
+			{ roles: options.accountability.roles },
+			async (fields) => {
+				const rolesService = new RolesService(context);
 
-		contextData['$CURRENT_ROLES'] = await rolesService.readMany(options.accountability.roles, {
-			fields: Array.from(permissionContext.$CURRENT_ROLES!),
-		});
+				return await rolesService.readMany(options.accountability.roles, {
+					fields,
+				});
+			},
+		);
 	}
 
 	if (options.policies.length > 0 && (permissionContext.$CURRENT_POLICIES?.size ?? 0) > 0) {
-		const policiesService = new PoliciesService(context);
+		contextData['$CURRENT_POLICIES'] = await fetchContextData(
+			'$CURRENT_POLICIES',
+			permissionContext,
+			{ policies: options.policies },
+			async (fields) => {
+				const policiesService = new PoliciesService(context);
 
-		contextData['$CURRENT_POLICIES'] = await policiesService.readMany(options.policies, {
-			fields: Array.from(permissionContext.$CURRENT_POLICIES!),
-		});
+				return await policiesService.readMany(options.policies, {
+					fields,
+				});
+			},
+		);
 	}
 
 	return contextData;
 }
+
+async function fetchContextData(
+	key: keyof RequiredPermissionContext,
+	permissionContext: RequiredPermissionContext,
+	cacheContext: Record<string, any>,
+	fetch: (fields: string[]) => Promise<Record<string, any>>,
+) {
+	const { cache } = getCache();
+	const env = useEnv();
+
+	const fields = Array.from(permissionContext[key]!);
+	const cacheKey = cache ? `filter-context-${key.slice(1)}-${getSimpleHash(JSON.stringify(cacheContext))}` : '';
```

Comment:
This cache key is simple and has a high chance of collision. Including the `fields` in the cacheKey should be better but more testing is needed to confirm. 🤔 

On a side note, this `cache` would not be purged if `CACHE_AUTO_PURGE=false`.

Reply:
Oh! The fields were meant to be included! Good catch!

Reply:
And yes, that cache would not be cleared, but then also the request level cache (which is the same for that matter) would not be cleared either. So you already have stale data.